walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,13 +11,140 @@
+    Collection<TimeRange> availableTimesBothAttendees = intersection(
+      availableTimesOptionals,
+      availableTimes
+    );
+
+    // Return time if any which works for both required and optional attendees
+    if (!availableTimesBothAttendees.isEmpty()) {
+      return availableTimesBothAttendees;
+    }
+
+    return availableTimes;
+  }
+
+  /**
+   * Finds avaliable times for the given atteendee list, input
+   * collection and event duration
+   */
+  private Collection<TimeRange> getAvailableTimes(Collection<String> attendees, Collection<Event> events, int duration) {
+    // TODO: Make this method more optimal
+    Collection<TimeRange> availableTimes = new ArrayList<>();
+
+    // possible start and end times of the event
+    int start = 0;
+    int end = start + duration;
+
+    // start and end time of the free period when event can occur
+    int availablePeriodStart = -1;
+    int availablePeriodEnd = -1;
+
+    // Consider every possible time period and see if it works for attendees
+    while (end <= TimeRange.END_OF_DAY) {
+      TimeRange currentTimeRange = TimeRange.fromStartDuration(start, duration);
+      boolean works = true;
+
+      // Comparing possible time range to the events to see if there is any overlaps
+      for (Event event : events) {
+        TimeRange eventTimeRange = event.getWhen();
+        if (currentTimeRange.overlaps(eventTimeRange)) {
+          Collection<String> eventAttendees = event.getAttendees();
+          if (!Collections.disjoint(attendees, eventAttendees)) {
+            works = false;
+          }
+        }
+
+        if (!works) break;
+      }
+
+      if (works) {
+        if (availablePeriodEnd == -1) {
+          availablePeriodStart = start;
+        }
+
+        availablePeriodEnd = end;
+
+        if (end == TimeRange.END_OF_DAY) {
+          TimeRange availableTime = TimeRange.fromStartEnd(
+            availablePeriodStart,
+            availablePeriodEnd,
+            true
+          );
+          availableTimes.add(availableTime);
+        }
+      } else {
+        if (availablePeriodEnd != -1) {
+          TimeRange availableTime = TimeRange.fromStartEnd(
+            availablePeriodStart,
+            availablePeriodEnd,
+            false
+          );
+          availableTimes.add(availableTime);
+        }
+
+        availablePeriodStart = -1;
+        availablePeriodEnd = -1;
+      }
+      start++;
+      end++;
+    }
+
+    return availableTimes;
+  }
+
+  /**
+   * Finds intersection between elements of two lists of
+   * TimeRange and returns as a list
+   */
+  private List<TimeRange> intersection(Collection<TimeRange> list1, Collection<TimeRange> list2) {
+    List<TimeRange> intersection = new ArrayList<>();
+
+    for (TimeRange el1 : list1) {
+      for (TimeRange el2 : list2) {
+        if (el1.contains(el2)) {
+          intersection.add(el2);
+        } else if (el2.contains(el1)) {
+          intersection.add(el1);
+        }
```

Comment:
Does this need to also cover the case of partial overlap? 

e.g. el1 = [1, 5], el2 = [3, 7]  => add [3, 5]

Reply:
I think another way to do this is something like:

```
int overlapStart = Math.max(el1.start(), el2.start());
int overlapEnd = Math.min(el1.end(), el2.end());
if (overlapStart < overlapEnd) {
  intersection.add(new TimeRange(overlapStart, overlapStart - overlapEnd);
}
```

Reply:
Oh you are right! overlapStart and overlapEnd also need to satisfy (overlapEnd - overlapStart >= duration), because time frame needs to be at least event's length.